gnovm/pkg/gnolang/type_check.go

```diff
@@ -997,6 +998,40 @@ func assertValidAssignLhs(store Store, last BlockNode, lx Expr) {
 	}
 }
 
+func assertValidAssignRhs(store Store, last BlockNode, n Node) {
+	var exps []Expr
+	switch x := n.(type) {
+	case *ValueDecl:
+		exps = x.Values
+	case *AssignStmt:
+		exps = x.Rhs
+	default:
+		panic(fmt.Sprintf("unexpected node type %T", n))
+	}
+
+	for _, exp := range exps {
+		tt := evalStaticTypeOfRaw(store, last, exp)
+		if tt == nil {
+			switch x := n.(type) {
+			case *ValueDecl:
+				if x.Type != nil {
+					continue
+				}
+				panic("use of untyped nil in variable declaration")
+			case *AssignStmt:
+				if x.Op != DEFINE {
```

Comment:
think this:
```go
package main

func main() {
	var a = 1
	a = nil
	println(a)
}
```

Reply:
same answer I think. 
https://github.com/gnolang/gno/pull/2896#discussion_r1827322486

Reply:
currently, `a := nil` and `a = nil` is handled in different places, the first one is handled here, and the latter one is handled in #3068, which is a bit weird.

is `if tt == nil {...}` only about handling a nil RHS? is so, can we defer this check to #3068, as nil is actually a valid RHS of an assignment in some cases. WDYT?

Reply:
I feel your original solution is close but can be optimized:
1. we can check if the expr on the right hand is XXXTypeExpr, if true, panic;
2. else if it is selectorExpr,  evalStaticType, check if it is TypeType;
3. else all good.
... am I wrong?

Reply:
> currently, a := nil and a = nil is handled in different places, the first one is handled here, and the latter one is handled in https://github.com/gnolang/gno/pull/3068, which is a bit weird.
> 
> is if tt == nil {...} only about handling a nil RHS? if so, can we defer this check to https://github.com/gnolang/gno/pull/3068 (for the compatibility assertion), as nil is actually a valid RHS of an assignment in some cases. WDYT?

I think `a := nil` and `a = nil` are not equivalent. In the first case, there is no defined type, rendering the assignment invalid, whereas in the second case, `a` has a type of `int`, allowing us to verify that the type of `a` is compatible with the value on the right side.

Reply:
> I think a := nil and a = nil are not equivalent. In the first case, there is no defined type, rendering the assignment invalid, whereas in the second case, a has a type of int, allowing us to verify that the type of a is compatible with the value on the right side.

this makes sense to me, unless a comment is needed for clarification, I suggest.

>I think the solutions are effectively the same. I've removed the if condition because XXXTypeExpr will already return TypeType when evaluated with evalStaticTypeOfRaw. So, in both cases, we end up checking for the same condition. Let me know if you see it differently!

yes sure, just think evalStaticTypeOfRaw will be expensive than a simple type assertion, is it?

Thank you for your explanation! I'll approve this soon after your feedback.

Reply:
Yes I think `evalStaticTypeOfRaw` performed a type assertion for a type that we can determine without evaluation, such as `constTypeExpt` and `ConstExpr`.